GCEWindowsAgent/serial_listener.go

```diff
@@ -0,0 +1,196 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os/exec"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+type serialMessage interface{}
+
+// Must match json format
+type prepareForSnapshotRequest struct {
+	Signature    string
+	Version      int
+	Operation_id int
+}
+
+type prepareForSnapshotResponse struct {
+	Signature    string
+	Version      int
+	Rc           int
+	Operation_id int
+}
+
+type resumePostSnapshotRequest struct {
+	Signature    string
+	Version      int
+	Operation_id int
+}
+
+type resumePostSnapshotResponse struct {
+	Signature    string
+	Version      int
+	Rc           int
+	Operation_id int
+}
+
+const preSnapshotDir = "/etc/google.d/disks/pre-snapshot/"
+const postSnapshotDir = "/etc/google.d/disks/post-snapshot/"
+
+// TODO I don't know what this should be
+const scriptTimeout = time.Second * 60
+
+// TODO I don't know what this should be
+const serialPort = "5"
+
+var (
+	currentMsg []byte
+)
+
+func clearCurrentMsg() {
+	currentMsg = nil
+}
+
+func parseMessage() serialMessage {
+	defer clearCurrentMsg()
+	var msgJson map[string]interface{}
+	err := json.Unmarshal(currentMsg, &msgJson)
+	if err != nil {
+		logger.Fatalf(err.Error())
+	}
+	sig, ok := msgJson["signature"]
+	if !ok {
+		logger.Errorf("message does not have signature")
+	}
+
+	switch sig {
+	case "PREREQ":
+		msg := prepareForSnapshotRequest{}
+		err := json.Unmarshal(currentMsg, &msg)
+		if err != nil {
+			logger.Fatalf(err.Error())
+		}
+		return msg
+	case "POSTREQ":
+		msg := resumePostSnapshotRequest{}
+		err := json.Unmarshal(currentMsg, &msg)
+		if err != nil {
+			logger.Fatalf(err.Error())
+		}
+		return msg
+	default:
+		logger.Errorf("unable to parse signature %s", sig)
+		return nil
+	}
+}
+
+func runScripts(dir string) int {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		logger.Errorf("unable to read directory %s", dir)
+	}
+	for _, f := range files {
+		fullpath := path.Join(dir, f.Name())
+		logger.Infof("running script at %s", fullpath)
+		cmd := exec.Command(fullpath)
+
+		cmd.Start()
+
+		done := make(chan error)
+		go func() { done <- cmd.Wait() }()
+
+		// Start timer
+		timeout := time.After(scriptTimeout)
+
+		select {
+		case <-timeout:
+			logger.Warningf("timeout while running %s", fullpath)
+			// Exit code of the timeout command when it times out
+			return 124
+		case err := <-done:
+			if err != nil {
+				exiterr := err.(*exec.ExitError)
+				status := exiterr.Sys().(syscall.WaitStatus)
+				logger.Warningf("script at %s return with exit code %d", fullpath, status.ExitStatus())
+				return status.ExitStatus()
+			}
+		}
+	}
+	return 0
+}
+
+func handlePrepareForSnapshotRequest(req prepareForSnapshotRequest) {
+	code := runScripts(preSnapshotDir)
```

Comment:
should any of the request attributes be handed to `runScripts` for logging and debugging purposes?